sys/exec.c

```diff
@@ -0,0 +1,185 @@
+#include <common.h>
+#include <stdc.h>
+#include <elf/mips_elf.h>
+#include <vm_map.h>
+#include <vm_pager.h>
+#include <thread.h>
+
+extern uint8_t _binary_prog_uelf_start[];
+extern uint8_t _binary_prog_uelf_size[];
+extern uint8_t _binary_prog_uelf_end[];
+
+int exec(){
+    uint8_t* elf_image = _binary_prog_uelf_start;
+    size_t elf_size = (size_t)(void*)_binary_prog_uelf_size;
+
+    kprintf("[exec] User ELF size: %ld\n", elf_size);
+
+    if(elf_size < sizeof(Elf32_Ehdr)){
+        kprintf("[exec] Exec failed: ELF file is too small to contain a valid header\n");
+        return -1;
+    }
+
+    const uint8_t header_size = 64;
+    const uint8_t stride = 8;
+    kprintf("[exec] User ELF header dump:\n");
+    for(int8_t i = 0; i < header_size; i++){
+        if(i % stride == 0) kprintf("[exec]  ");
+        uint8_t byte = elf_image[i];
+        kprintf(" %02x", byte);
+        if(i % stride == 7) kprintf("\n");
+    }
+
+    const Elf32_Ehdr* eh = (Elf32_Ehdr*)elf_image;
+
+    // Start by determining the validity of the elf file.
+
+    // First, check for the magic header.
+    if(eh->e_ident[EI_MAG0] != ELFMAG0 ||
+       eh->e_ident[EI_MAG1] != ELFMAG1 ||
+       eh->e_ident[EI_MAG2] != ELFMAG2 ||
+       eh->e_ident[EI_MAG3] != ELFMAG3){
+        kprintf("[exec] Exec failed: Incorrect ELF magic number\n");
+        return -1;
+    }
+    // Check ELF class
+    if(eh->e_ident[EI_CLASS] != ELFCLASS32){
+        kprintf("[exec] Exec failed: Unsupported ELF class (!= ELF32)\n");
+        return -1;
```

Comment:
That seems fine. However, I suggest we implement `errno` in a separate branch. It will probably be useful in many functions throughout the entire source.